### ISSUE TITLE
Components can't safely import React

### ISSUE THREAD

**Issue by ericelliott** [stampit-org/react-stampit#13]
I just noticed a pretty serious flaw in react-stampit:

**Component source files should never import React**, because that could lead to multiple instances and multiple versions of React in the same project -- a situation that causes bugs. Instead, it should be passed into the stamp when an instance is instantiated (dependency injection). You'll need to work some magic to expose the React.PropTypes for the react-stamp to use.

A fix for this issue will be a breaking change, and any users who've managed to find this already should upgrade immediately to the fixed version to prevent bugs in their applications.

Example from [react-hello](https://github.com/stampit-org/react-hello):

``` js
import React from 'react'; // we can't do this...
import stampit from 'react-stampit';

const {
  string,
  func
} = React.PropTypes;
```

**Comment by troutowicz** [comment]
Is this not what components do with `React.createClass`? I wasn't aware of bugs resulting from this.

**Comment by ericelliott** [comment]
No, not if the mixins are well designed. **Mixins should never require their own copy of React**, for just this reason. One of the main reasons that composing real components is a better idea than using Mixins.

**Anything that would cause two React instances should be an anti-pattern** just because the browser may have to load React twice, so even if it didn't actually cause bugs in software which is [one of the most common issues new users face](https://medium.com/@dan_abramov/two-weird-tricks-that-fix-react-7cf9bbdef375), this should still be considered a bug.

**Comment by troutowicz** [comment]
Thanks for the link. I'll work on the change. :+1:

**Comment by troutowicz** [comment]
So this is one way we can fix it.
- React.PropTypes
  
  Export a React object stub from react-stampit that contains an empty object property `PropTypes`. The developer would import this in components that require it. Then the real React lib would take priority when passed as an argument to the factory.
- API

``` js
test('stampit(props)(React)', (t) => {
  t.plan(1);

  const stampFactory = stampit({
    propTypes: {
      foo: React.PropTypes.string
    },

    render() {
      return null;
    }
  });
  const stamp = stampFactory(React);

  t.ok(
    stampit.isStamp(stamp),
    'should return a stamp'
  );
});
```

We could maybe create a convenience method on the stamp factory for some simplicity.

``` js
const stamp = stampit({
  propTypes: {
    foo: React.PropTypes.string
  },

  render() {
    return null;
  }
}).inject(React);
```

**Comment by ericelliott** [comment]
I don't understand those examples. Both depend on `React` directly in the definition of the stamp, and then pass `React` into the stamp during instantiation.

The point of this issue is that `React` won't be available to depend on when the user is defining the stamp.

**Comment by troutowicz** [comment]
My example is partially wrong. The `.inject` method is not needed and actually doesn't make sense. Let me split things up.

Edit: This actually won't work either, need to rethink how to handle `React.PropTypes`...

**component.jsx**

``` js
import { React } from 'react-stampit';

export default stampit({
  propTypes: {
    foo: React.PropTypes.string
  },

  render() {
    return null;
  }
});
```

**app-client.jsx**

``` js
import React from 'react';
import componentFactory from './component';

const Component = componentFactory(React);

React.render(
    <Component />,
    document.getElementById('content')
);
```

**Comment by ericelliott** [comment]
That makes more sense, but wouldn't this be more clear?

``` js
import { PropTypes } from 'react-stampit';

export default stampit({
  propTypes: {
    foo: PropTypes.string
  },

  render() {
    return null;
  }
});
```

Also, PropTypes could not be "empty" - it would actually have to contain the full React.PropTypes object structure so that things like `PropTypes.func.required` don't throw exceptions.

**Comment by ericelliott** [comment]
This wouldn't be such a headache if React itself was a bit more modular. `PropTypes` should really be its own module so we could just do something like `import PropTypes from 'react-proptypes';` and we wouldn't have to worry about multiple versions of the behavioral bits of React floating around in our apps...

**Comment by troutowicz** [comment]
It does, but I was trying to keep things familiar. Either way, it doesn't matter because that approach will fail. The properties will be `undefined`.

**Comment by troutowicz** [comment]
> PropTypes should really be its own module

This may be what we need to create.

**Comment by ericelliott** [comment]
Who says the properties have to be `undefined`? As I mentioned in my comment above, you'd have to supply the full `React.PropTypes` object structure for this to work. =)

**Comment by troutowicz** [comment]
Ah, missed you saying that, our posts came in a bit out of order. ;)

**Comment by ericelliott** [comment]
:+1:

**Comment by troutowicz** [comment]
@ericelliott I feel I have to say one last thing here before continuing this breaking change. The only time multiple instances of React would occur is if an app, which has React as a dependency, has dependencies that also have their own React dependency, right? So is the concern here that components/mixins would not be safely distributable if React was a dependency? That's the only reason I can think of.

**Comment by ericelliott** [comment]
Not the only reason, because many apps (including mine) source React from CDN (likely already cached by browsers).

I believe both Browserify and Webpack automatically dedupe dependencies, so this should theoretically only be a dependency problem when a component or mixin imports a conflicting version of React, which is unfortunately a common scenario that has led to many GitHub issues and StackOverflow questions.

**Comment by troutowicz** [comment]
Looks like someone has made a module for [PropTypes](https://github.com/aackerman/PropTypes).

**Comment by troutowicz** [comment]
@ericelliott Created a [refactor](https://github.com/stampit-org/react-stampit/tree/refactor) branch. I need to work through all the tests, and update docs. I did update the example in the README though.

**Comment by ericelliott** [comment]
Nice!

**Comment by troutowicz** [comment]
@ericelliott I just realized PR #14 is largely not needed. Components can be made factories by the developer, it does not and should not be the libraries responsibility IMO. If the library handles this, then components that have a render function returning something other than null will error because `React` is not defined inside the component. (`React.createElement`)

Creating a factory component without the PR

``` js
import { PropTypes } from 'prop-types';

export default React => stampit(React, {
  propTypes: {
    foo: PropTypes.string
  },

  render() {
    return null;
  }
});
```

Creating a factory component with the PR

``` js
import { PropTypes } from 'prop-types';

export default React => stampit({
  propTypes: {
    foo: PropTypes.string
  },

  render() {
    return null;
  }
})(React);
```

The latter results in a extra function call and a breaking change in the API.

**Comment by ericelliott** [comment]
This looks like the cleanest solution to the problem.

``` js
import { PropTypes } from 'prop-types';

export default React => stampit(React, {
  propTypes: {
    foo: PropTypes.string
  },

  render() {
    return null;
  }
});
```

Want to update the docs to clarify this?

**Comment by troutowicz** [comment]
Yeah, I'm going to take any useful changes from this PR and update the docs.

**Comment by ericelliott** [comment]
:sunglasses:

**Comment by troutowicz** [comment]
Closed, see #15
